### PR TITLE
Add support for specifying fixed RTK base position

### PR DESCRIPTION
**Pull request by DonLakeFlyer**
This allows you to specify a fixed RTK base position and skip survey-in. If you have a valid survey-in "Save Current Base Position" will enable allowing you to copy the values from the survey-in to setttings.

Replacement for #6868. Related to #6621.

screen shot 2018-10-01 at 9 49 36 am

### CODE REVIEW

**Comment by DonLakeFlyer** [comment]
@bkueng In my testing this seems to be working. But I am not RTK expert so could definitely use some additional testing.

**Comment by DonLakeFlyer** [comment]
@hamishwillee Do you think you could add the doc for this into the Daily Build page?

File: src/GPS/GPSProvider.cc [hunk 1]
```diff
@@ -70,7 +85,11 @@ void GPSProvider::run()
             gpsDriver = new GPSDriverUBX(GPSDriverUBX::Interface::UART, &callbackEntry, this, &_reportGpsPos, _pReportSatInfo);
             baudrate = 0; // auto-configure
         }
-        gpsDriver->setSurveyInSpecs(_surveyInAccMeters * 10000, _surveryInDurationSecs);
+        gpsDriver->setSurveyInSpecs(_surveyInAccMeters * 10000.0f, _surveryInDurationSecs);
+
+        if (_useFixedBaseLoction) {
+            gpsDriver->setBasePosition(_fixedBaseLatitude, _fixedBaseLongitude, _fixedBaseAltitudeMeters * 10000.0f, _fixedBaseAccuracyMeters * 10000.0f);
```
Comment (bkueng) [inline src/GPS/GPSProvider.cc]: Altitude is given in meters, and accuracy in mm (one 0 too much):
`_fixedBaseAltitudeMeters, _fixedBaseAccuracyMeters * 1000.0f`

File: src/QGCApplication.cc [hunk 2]
```diff
@@ -197,7 +199,7 @@ QGCApplication::QGCApplication(int &argc, char* argv[], bool unitTesting)
             QMessageBox msgBox;
             msgBox.setInformativeText(tr("You are running %1 as root. "
                                          "You should not do this since it will cause other issues with %1. "
-                                         "%1 will now exit. "
+                                         "%1 will now exiQGroundControl.gpsRtkt. "
```
Comment (bkueng) [inline src/QGCApplication.cc]: typo?

File: src/Settings/RTK.SettingsGroup.json [hunk 3]
```diff
@@ -20,5 +20,55 @@
...
+},
+{
+    "name":                 "FixedBasePositionLatitude",
+    "shortDescription":     "Base Position Latitude",
+    "longDescription":      "Defines the latitude of the fixed RTK base position.",
+    "type":                 "double",
+    "defaultValue":         0,
+    "min":                  -90,
+    "max":                  90,
+    "decimalPlaces":        7,
```
Comment (bkueng) [inline src/Settings/RTK.SettingsGroup.json]: Can you increase this to 9? That's what's used internally.

File: src/Vehicle/GPSRTKFact.json [hunk 4]
```diff
@@ -2,35 +2,62 @@
...
     "decimalPlaces":    1,
-    "units":            "m"
+    "units":            "m",
+    "default":          null
+},
+{
+    "name":             "currentLatitude",
+    "shortDescription": "Current Survey-In Latitude",
+    "type":             "double",
+    "decimalPlaces":    7,
```
Comment (bkueng) [inline src/Vehicle/GPSRTKFact.json]: Also: 9

File: src/GPS/GPSProvider.cc (hunk 1 above)
Comment (DonLakeFlyer) [reply src/GPS/GPSProvider.cc]: Oops, thanks.

Yeah I was thinking about disabling the other ones. Let me look at that.

Other question: The field editors show a maximum number of decimal places. Although internally the full value is stored, visibly only 7 digits are shown after the decimal. What do you think the visible precision should me. Is 7 too few?

File: src/QGCApplication.cc (hunk 2 above)
Comment (DonLakeFlyer) [reply src/QGCApplication.cc]: Yup.

File: src/Settings/RTK.SettingsGroup.json (hunk 3 above)
Comment (DonLakeFlyer) [reply src/Settings/RTK.SettingsGroup.json]: Ah, thanks. I missed this before I posted my above decimal place question.

File: src/GPS/GPSProvider.cc (hunk 1 above)
Comment (DonLakeFlyer) [reply src/GPS/GPSProvider.cc]: Disregard decimal place dup question. Already answered,.

**Comment by DonLakeFlyer** [comment]
screen shot 2018-10-02 at 12 58 54 pm

**Comment by DonLakeFlyer** [comment]
@mkunzBGC Do you have time to test once I merge in?

**Comment by DonLakeFlyer** [comment]
@bkueng Is altitude and accuracy ok with 2 digit precision?

**Comment by mkunzBGC** [comment]
@DonLakeFlyer Yes, I can test once a daily build is available.

**Comment by DonLakeFlyer** [comment]
> he RTK icon will only show up when a vehicle is connected. Can you change that to always show up indepenent from a vehicle?

Yup, known problem (there an issue for it somewhere). The way the toolbar works it's a non-trivial change though.

**Comment by DonLakeFlyer** [comment]
@mkunzBGC New daily builds with this should be available in about an hour.

**Comment by mkunzBGC** [comment]
@DonLakeFlyer Today I tested QGroundControl Development HEAD:defdcdc41 2018-10-04 14:37:54 -0700, which includes the option to save and restore the base station position. My setup consisted of a Here+ RTK base and rover module, both running the 1.40 firmware. The rover was connected to a 3DR Pixhawk running ArduCopter 3.5.5. Rover and base antenna were set up under open sky, elevated 2 and ~1.2 m from the ground, respectively. They were statically mounted at a distance of ~5m horizontally and ~1.5m vertically. The base station saw 20 satellites all the time, the rover used 15-19.

During all testing I waited until RTK fixed mode before taking coordinates. I measured the position of the rover three times, restarting the rover and QGroundControl in between. Saving and restoring the base position worked as expected. The positions are within 2.5cm horizontally and 2 cm vertically to each other, which is a pretty good result.

However, the determined base station altitude was strange in that it differed from the rover by more than 40m (268m for the base, 223.4m for the rover). I am at 50.909°N and 11.565°E if that matters.

One suggestion: When saving the base position, several messages pop up: "Change of 'Base Position [Latitude/Longitude/Altitude/Accuracy]' value requires restart of QGroundControl to take effect". I think this is not necessary until the "Use specified base position" radio button is activated.

Thanks for adding this useful feature!

**Comment by DonLakeFlyer** [comment]
@mkunzBGC Thanks fo much for testing.
@bkueng Any ideas on the altitude thing?

**Comment by mkunzBGC** [comment]
@DonLakeFlyer The difference in altitude is likely related to the conversion between ellipsoid to geoid altitude as explained [here](http://www.gpspassion.com/forumsen/topic.asp?TOPIC_ID=10915). The NMEA stream from a different GPS receiver, captured at a location few kilometers away, indicates that the reference ellipsoid is 47.5 m higher than the geoid here. This fits nicely to the difference I saw in the base station altitude.

I imagine that the ublox chipset needs the ellipsoid altitude, but for the user interface it would be desirable to work with geoid altitude. If the base station outputs the altitude correction then QGroundControl could do a conversion that is invisible to the user. If this info is not available I would suggest a hint to the user right next to the field were the base station position is indicated.

**Comment by bkueng** [comment]
@mkunzBGC Sorry I dropped this. You are correct: the vehicle reports the mean sealevel altitude, whereas the RTK base station outputs and uses the ellipsoid altitude.
Either solution is fine with me: add a remark to the UI or do the conversion.

**Comment by DonLakeFlyer** [comment]
So "Base Position Altitude" as shown in image in previous command is WGS84. But what other value is AMSL from where specifically?

**Comment by mkunzBGC** [comment]
"Base Position Altitude" is ellipsoid altitude, i.e. uses the coordinate system where the surface of the WGS84 reference ellipsoid is at altitude 0. The reference ellipsoid is a very simple representation of the earth and therefore deviates quite a bit from what we call mean sea level. GPS receivers include a lookup table that allows them to convert from ellipsoid altitude to altitude above mean sea level.  The NMEA message GGA includes this converted altitude in field 9 (alt) *and* the applied offset in field 11 (sep).
I guess that the lookup tables are different from receiver model to receiver model and could even change with firmware updates. However, I would assume that they are the same for base and rover. As the correction is (afaik) always applied to the rover, I would expect the most consistent results if they are also applied to the base station altitude.
@DonLakeFlyer Is the info from GGA field 11 available at the time when the user would save the base position, i.e. does the RTK base send this info? If not, use ellipsoid altitude and inform the user, e.g. "This is altitude above the WGS84 ellipsoid, not altitude above mean sea level"
If the info is available, the correction could be applied to the altitude. However, the base station internally uses ellipsoid altitude, so the question is how to load a saved position. I see three options:

1. Wait until the base has a good GPS fix so that it tells you the correct separation, allowing the conversion from the saved altitude to ellipsoid altitude.
2. Save the separation together with the base station coordinates. It is not enough to save this "silently" - it should be another field where the user can enter values.
3. Define the the position of the base station by lat, lon and ellipsoid altitude, but inform the user in a read-only way about the altitude above mean sea level.

I would prefer option 3.

**Comment by bkueng** [comment]
Option 3 sounds good to me too. If we start to add offsets we introduce inaccuracies.
A fixed lookup-table would work too, but I'm not sure if there's a well-defined standard for it.

**Comment by DonLakeFlyer** [comment]
> Define the the position of the base station by lat, lon and ellipsoid altitude

I get that I think. It's just label the altitude of "Base Position Altitude" as WGS84. So text changes to something like "Base Position Alt (WGS84)"?

>  inform the user in a read-only way about the altitude above mean sea level.

Inform exactly where?

**Comment by mkunzBGC** [comment]
I made a mockup of an input mask that would imho be (1) intuitive (the user immediately knows which values need to be entered), (2) explicit (the user sees that two different altitude references exist and gets enough info to look up details if desired) and (3) error-proof (potential inconsistencies in altitude conversion do not influence the RTK solution because only the ellipsoid altitude is used):
inputmaskrtkbase3

**Comment by DonLakeFlyer** [comment]
I'd rather not have to get into WGS84 to AMSL conversion unless I really need to. Does that value come the the GPS as well?

**Comment by DonLakeFlyer** [comment]
So Lorenz is trying to figure out whether there is a WGS84 to AMSL conversion code available. At this point it appears to be ublox proprietary. Until then the best I can do is change the label on Base Position Altitude to state that it's WGS84.

**Comment by mkunzBGC** [comment]
@DonLakeFlyer Does the base station send GGA messages to the host computer? If so, you can just use the "sep" value in field 11 for the conversion from ellipsoid to asl altitude. If the base station does not provide this NMEA sentence, is there another way to get the separation used by the ublox chipset? It is an approximation, so calculating the separation in QGroundControl will likely be inconsistent to what the ublox chipset does for the rover altitude.

**Comment by DonLakeFlyer** [comment]
Those would be questions for @bkueng

**Comment by bkueng** [comment]
> @DonLakeFlyer Does the base station send GGA messages to the host computer?

It uses NAV-PVT, available in QGC from here: https://github.com/mavlink/qgroundcontrol/blob/master/src/GPS/GPSProvider.cc#L162. It contains both altitudes. However these altitudes are the current measurement, not based on the survey-in, and thus less accurate.

**Pull request merged**